README.md
# Pancake

<div align="center">
	<img src="https://github.com/Rabios/Pancake/blob/master/pancake.png" width="256", height="256"><br>
	<p><q>Lightweight, Fast, Easy-to-use HTML5 2D game framework!</q></p>
</div><br>
<br>

Pancake is successor of Cake game engine, Which can be found [here](https://github.com/steria773-archive/Cake), But no longer maintained!

Pancake inspired by LOVE, Pygame, And even AppGameKit and it shares some of their concepts...

## Features

- Shares concepts of most lovable frameworks like LOVE, Pygame, And AppGameKit!
- Easy and simple to setup!
- Free, Moddable, Open-Source, Cross-Platform!
- Fast, Very lightweight, And very thin in one file!
- Every part needed is written without dependencies! From OS detection to gamepad support...
- You can use `CanvasRenderingContext2D` or `WebGLRenderingContext` as graphics renderer!
- Collision detection physics!
- Audio based on HTMLAudio API!
- Video rendering support!
- Replay rewind support!
- Spritefonts and Sprites support!
- Have plugins can be used like WASM, FBInstant, And more...
- Written in plain JavaScript (Vanilla JavaScript) to support browsers that can't run modern JavaScript! (Internet Explorer for example)
- Supports [Haxe](https://haxe.org) via [Pancake.hx](https://github.com/Rabios/Pancake.hx)
- Using indexes (Slots), Which can be good for ordering sometimes, And gives easy control of game content!
- Optimized for js13kGames game jam!

Want to know a lot? See the [documentation](https://github.com/Rabios/Pancake/tree/master/docs/README.md)

## Examples

- Creating a game canvas

```javascript
// Create canvas with both width and height of 600, And set it's index to 0
pancake.canvas.create(600, 600, 0);

// Create context with index 0 and set to use canvas that has index 0
// And set graphics to use context with index 0
pancake.context.create(0, 0);
pancake.graphics.useContext(0);
```

- Drawing text

```javascript
// Create canvas with both width and height of 600, And set it's index to 0
pancake.canvas.create(600, 600, 0);

// Create context with index 0 and set to use canvas that has index 0
// And set graphics to use context with index 0
pancake.context.create(0, 0);
pancake.graphics.useContext(0);

// We can set font and it's size in pixels
pancake.graphics.setFont("Arial", 22);

// Draw text
pancake.graphics.text("Hello Pancake", 10, 10);
```

- Drawing shapes

```javascript
// Create canvas with both width and height of 600, And set it's index to 0
pancake.canvas.create(800, 800, 0);

// Create context with index 0 and set to use canvas that has index 0
// And set graphics to use context with index 0
pancake.context.create(0, 0);
pancake.graphics.useContext(0);

// Draw rectangle with red color
pancake.graphics.color(pancake.graphics.RGB(255, 0, 0));
pancake.graphics.rect(100, 100, 50, 50);

// Draw circle with radius of 50 and orange color
pancake.graphics.color(pancake.graphics.RGB(255, 165, 0));
pancake.graphics.circle(200, 200, 50);

// Draw line with width of 5 and yellow color
pancake.graphics.color(pancake.graphics.RGB(255, 255, 0));
pancake.graphics.line(100, 100, 200, 200, 5);

// Set drawing mode to stroke
pancake.graphics.mode = pancake.graphics.STROKE;

// Draw triangle with green color
pancake.graphics.color(pancake.graphics.RGB(0, 255, 0));
pancake.graphics.triangle(500, 50, 600, 50, 500, 150);

// Draw polygon with 6 sides and blue color
pancake.graphics.color(pancake.graphics.RGB(0, 0, 255));

pancake.graphics.polygon(100, 300, 50, 6);
```

- Playing Audio

```javascript
// When audio loaded, Play it directly
pancake.audio.play("game.mp3");

// Load audio file to index 0
pancake.audio.load("loaded.mp3", 0);

// Plays loaded.mp3
pancake.audio.playFromIndex(0);

// Pauses loaded.mp3
pancake.audio.pause(0);

// Resumes loaded.mp3
pancake.audio.playFromIndex(0);
```

- Saving variables

```javascript
// Save score to localStorage variable "score"
pancake.storage.save("score", Number(0));

// Load it,It gives score value which is 0
var score = Number(pancake.storage.load("score"));
```

- Physics + Game Loop

```javascript
var playercolor = pancake.graphics.RGB(0, 0, 255);

// Use a HTML <canvas> tags that has ID "canvas"
// And set index of the context that will use it to 0
pancake.context.use(document.getElementById("canvas"), 0);

// Create 2 rectangles objects
var rect1 = { x: 100, y: 100, w: 50, h: 50 };
var rect2 = { x: 125, y: 125, w: 50, h: 50 };

function game() {
    // Clear canvas
    pancake.graphics.clear();

    // Draw first rectangle
    pancake.graphics.color(playercolor);
    pancake.graphics.rect(rect1.x, rect1.y, rect1.w, rect1.h);
    
    // Draw second rectangle
    pancake.graphics.color(pancake.graphics.RGB(0, 255, 0));
    pancake.graphics.rect(rect2.x, rect2.y, rect2.w, rect2.h);

    // Check collision between both 2 rectangles
    // If true then change color of first rectangle
    // Else then reset it's color to the default color
    if (pancake.physics.checkCollisionRecs(rect1.x, rect1.y, rect1.w, rect1.h, rect2.x, rect2.y, rect2.w, rect2.h)) {
        playercolor = pancake.graphics.RGB(255, 0, 0);
    } else {
        playercolor = pancake.graphics.RGB(0, 0, 255);
    }   
}

var gameloop = pancake.timers.timer(game, 60); // Set frames per second to 60
```

- Keyboard + Game Loop

```javascript
function game() {
    // If key down
    if (pancake.input.keydown(pancake.input.key.X)) alert("Key down/press");

    // If key up
    if (pancake.input.keyup(pancake.input.key.C)) alert("Key up");

    // Needed to prevent input loop
    pancake.input.preventLoop();
}

var gameloop = pancake.timers.timer(game, 60); // Set frames per second to 60
```

> NOTE: More examples can be found in the [Examples folder](https://github.com/Rabios/Pancake/tree/master/examples)

## Build

Build using Python:

```batch
mkdir pancake
cd pancake
git clone https://github.com/Rabios/Pancake.git
python build.py
```

A folder named `build` will created (if not exist) in the repository folders, Containing `pancake.js`, Which is the full build.

> For more info about building Pancake see [here](https://github.com/Rabios/Pancake/blob/master/docs/build_system.md).

## License

```
MIT License

Copyright (c) 2020 - 2022 Rabia Alhaffar

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
```


docs/changelog.md
# Changelog

> NOTE: For each update, I recommend you to always take look at [API](https://github.com/Rabios/Pancake/blob/master/docs/api.md) for changes occur.

### v0.0.16 (Update 16)

Added `pancake.os.TIZEN` and Tizen interface for `pancake.game.close()` to support Tizen!

### v0.0.15 (Update 15)

1. Fixed gamepad input limits on UWP platform, Now gamepad input works %100!
2. Fixed alpha setting for `pancake.graphics.random.RGBA` and `pancake.graphics.RGBA` for `CanvasRenderingContext2D` backend!
3. Fixed analog naming (Too old naming mistake...)

### v0.0.14 (Update 14)

1. Added support and improvements for working with native stuff like NWJS, Electron, Phonegap, Cordova, WinJS, and UWP...
2. As gamepad functions not supported by Internet Explorer, You'll get null/false instead of being not exist in Pancake (That's to prove error-safe lib)
3. Done some improvements...
4. Added input-customizing variables like `pancake.input.gamepad_threshold`, `pancake.input.swipe_finish_x_limit`, `pancake.input.swipe_finish_y_limit`, `pancake.input.swipe_finish_time_limit`, `pancake.input.accelerometer_frequency`, Enjoy modding input sensitivity!

### v0.0.13 (Update 13)

1. Fixed WebGL renderer and rewrote with internal use of heavily modified version of [polygl.js](https://github.com/Rabios/polygl.js).
2. Added plugins like `pancake.wasm` for WebAssembly, And `pancake.clipboard` for Clipboard API support, And more...
3. Added input support for Accelerometer.
4. Released [Haxe port](https://github.com/Rabios/Pancake.hx) of Pancake!
5. Added Discord Server, Come join us [here](https://discord.gg/6asrfg6K)!
5. Merged all source in one folder with rename to the renderer, From `canvas` to `canvas2d` when building with Python.
6. Added `pancake.device.geoInfo()` and `pancake.device.language`.
7. Added `pancake.graphics.backend` variable so you can get type of rendering context used by Pancake.
8. Cursor now can be hidden with `pancake.input.showCursor()` and `pancake.input.hideCursor()`, For real as it's fixed!
9. Added input support for mouse wheel controls!
10. Added input support for TV remotes!
11. Added `pancake.graphics.setBackground()` to set canvas background with custom CSS style!
12. Removed 3rd parameter from `pancake.input.gamepadAnalogMoved`, So you just put 4th parameter after 2nd one and just enough (That's to make it easier...).
13. Added logo and icon for Pancake!
14. Added official website for Pancake!
15. Removed `pancake.debug` module...
16. Done a couple of optimizations...
17. `window.animate` became `pancake.timers.animate`, And updated polyfill of `requestAnimationFrame` and `cancelAnimationFrame`...
18. Fixed definition of `pancake.content` in `define.js`.
19. Changed build system, More info can be found [here!](https://github.com/Rabios/Pancake/blob/master/docs/api/build_system.md)
20. Changed functionality of `pancake.graphics.polygon`, And if you use code from previous versions you'll find breaks...
21. Rewrote drawing circles using the changed `pancake.graphics.polygon` function.
22. Fullscreen and Gamepad now supports old versions of WebKit.
23. Support for multi-fingering gestures, Even with swipe!
25. Images drawing got overloads and versions of same function, Like `pancake.graphics.drawImage` and `pancake.graphics.image` and `pancake.graphics.drawImageFromIndex` and `pancake.graphics.imageFromIndex`.
26. Edited gamepad input to use gamepad index as 1st parameter instead of being last in some functions.
27. Edited Audio module functions names to comply with Video module functions names...
28. Added `pancake.browser.OPERA_MINI` to detect if browser is Opera Mini!
29. Added `pancake.browser.is()` and can also be used as `pancake.os.is()` to detect if string is in userAgent.
30. Added `pancake.replay` for saving replays in game!
31. Removed antialiasing support from Pancake cause it makes textures/images drawing blurry.
32. Added `pancake.spritefont`, Which makes it possible to draw texts from images.
33. Renamed current `pancake.sprite` to `pancake.gif`, And made `pancake.sprite` for drawing sprite from cheatsheet!
34. `pancake.physics.getDistance` returns object contains `x` and `y` instead of being stored in variables!
35. Now offered support folder, It contains polyfills that can be useful if your browser doesn't support a feature by Pancake.
36. Readded `pancake.device.online()` function back after being removed since `v0.0.2`.
37. Added `pancake.xhr` for `XMLHttpRequest` Support!
38. Added `pancake.graphics.fit()` function, It lets game canvas fits entire window!
39. Renamed `pancake.browser.java` to `pancake.browser.support.JAVA`
40. Added scissor test for both renderers, Use them via `pancake.graphics.beginScissor` and `pancake.graphics.endScissor`
41. Remove some functions for compatibility and due to being low used, These are removed functions:

```js
pancake.graphics.setLineWidth
pancake.graphics.shadow
pancake.graphics.ellipse
pancake.graphics.setTextAlignment
pancake.graphics.gradientRect
pancake.script.eval					// Use window.eval instead
pancake.input.keypress				// Use pancake.input.keydown, or pancake.input.keyup instead for not repeat key press in game-loop
```

42. Renamed some Physics functions, It's better to take look [here!](https://github.com/Rabios/Pancake/blob/master/docs/api.md#pancakephysics)
43. Redesigned collision detection with canvas sides, So it uses object instead!
44. Added new collision detection functions, Following were added:

```js
pancake.physics.checkCollisionCirclePoint
pancake.physics.checkCollisionRectLine
pancake.physics.checkCollisionRectPoint
pancake.physics.checkCollisionLines
pancake.physics.checkCollisionLinePoint
pancake.physics.checkCollisionPoints
pancake.physics.checkCollisionTriangles
pancake.physics.checkCollisionTriangleRect
pancake.physics.checkCollisionTriangleLine
pancake.physics.checkCollisionTrianglePoint

src/input.js
p.i = {};
p.i.latest_key_down = -1;
p.i.latest_key_up = -1;
p.i.latest_mouse_button_down = -1;
p.i.latest_mouse_button_up = -1;
p.i.click = !1;
p.i.touchdown = !1;
p.i.tap = !1;
p.i.mouse_x = 0;
p.i.mouse_y = 0;
p.i.touch_x = 0;
p.i.touch_y = 0;
p.i.wheel_x = 0;
p.i.wheel_y = 0;
p.i.wheel_z = 0;
p.i.accel_x = 0;
p.i.accel_y = 0;
p.i.accel_z = 0;
p.i.wheel_up = !1;
p.i.wheel_down = !1;
p.i.wheel_left = !1;
p.i.wheel_right = !1;
p.i.swipe_start_time = 0;
p.i.swipe_start_x = 0;
p.i.swipe_start_y = 0;
p.i.swipe_finish_x = 0;
p.i.swipe_finish_y = 0;
p.i.swipe_finish_time = 0;
p.i.swipe_finish_time_limit = 1000;
p.i.swipe_finish_x_limit = { from: 100, to: 300 };
p.i.swipe_finish_y_limit = { from: 300, to: 300 };
p.i.gamepad_threshold = 0.1;
p.i.accelerometer_frequency = 60;
p.i.swipe_direction = "";
p.i.accelerometer = u;
p.i.touches = [];

if (w.Accelerometer) {
    p.i.accelerometer = new Accelerometer({ frequency: p.i.accelerometer_frequency });
    p.i.accelerometer.onreading = function(acl) {
        p.i.accel_x = acl.x;
        p.i.accel_y = acl.y;
        p.i.accel_z = acl.z;
    };
    p.i.accelerometer.start();
}

w.onmousedown = w.onmspointerdown = w.onpointerdown = function(e) {
    p.i.swipe_start_x = ex(e);
    p.i.swipe_start_y = ey(e);
    p.i.swipe_start_time = da();
    p.i.latest_mouse_button_down = e.button;
    p.i.click = !1;
};

w.onmouseup = w.onmspointerup = w.onpointerup = function(e) {
    p.i.mouse_x = ex(e);
    p.i.mouse_y = ey(e);
    p.i.swipe_finish_x = p.i.mouse_x - p.i.swipe_start_x;
    p.i.swipe_finish_y = p.i.mouse_y - p.i.swipe_start_y;
    p.i.swipe_finish_time = da() - p.i.swipe_start_time;
    if (p.i.swipe_finish_time <= p.i.swipe_finish_time_limit) {
        f = m.abs;
        if (f(p.i.swipe_finish_x) >= p.i.swipe_finish_x_limit.from && f(p.i.swipe_finish_y) <= p.i.swipe_finish_y_limit.to) {
            if (p.i.swipe_finish_x < 0) p.i.swipe_direction = "LEFT";
            else p.i.swipe_direction = "RIGHT";
        }
        else if (f(p.i.swipe_finish_y) >= p.i.swipe_finish_y_limit.from && f(p.i.swipe_finish_x) <= p.i.swipe_finish_x_limit.to) {
            if (p.i.swipe_finish_y < 0) p.i.swipe_direction = "UP";
            else p.i.swipe_direction = "DOWN";
        }
    }
    p.i.latest_mouse_button_up = e.button;
    p.i.click = !1;
};

w.onmousemove = w.onmspointermove = w.onpointermove = function(e) {
    p.i.mouse_x = ex(e);
    p.i.mouse_y = ey(e);
};

w.ontouchstart = function(e) {
    q = e.changedTouches;
    for (var i = 0; i < q.length; i++) {
        p.i.touches[i] = {
            x: ex(q[i]),
            y: ey(q[i]),
            swipe_start_x: q[i].pageX,
            swipe_start_y: q[i].pageY,
            swipe_start_time: da(),
            swipe_direction: "",
            tap: !0,
            touchdown: !1,
        };
    }
    v = p.i.touches[0];
    p.i.touch_x = v.x;
    p.i.touch_y = v.y;
    p.i.tap = !0;
    e.preventDefault();
};

w.ontouchmove = function(e) {
    q = e.changedTouches;
    for (var i = 0; i < q.length; i++) {
        t = p.i.touches[i];
        t.x = ex(q[i]);
        t.y = ey(q[i]);
        t.touchdown = !0;
    }
    v = p.i.touches[0];
    p.i.touch_x = v.x
    p.i.touch_y = v.Y;
    p.i.touchdown = !0;
    e.preventDefault();
};

w.ontouchend = function(e) {
    q = e.changedTouches;
    for (var i = 0; i < q.length; i++) {
        t = p.i.touches[i];
        t.x = ex(q[i]);
        t.y = ey(q[i]);
        t.swipe_finish_x = q[i].pageX - t.swipe_start_x;
        t.swipe_finish_y = q[i].pageY - t.swipe_start_y;
        t.swipe_finish_time = da() - t.swipe_start_time;
        t.tap = !0;
        t.touchdown = !1;
        if (t.swipe_finish_time <= p.i.swipe_finish_time_limit) {
            f = m.abs;
            if (f(t.swipe_finish_x) >= p.i.swipe_finish_x_limit.from && f(t.swipe_finish_y) <= p.i.swipe_finish_y_limit.to) {
                if (t.swipe_finish_x < 0) t.swipe_direction = "LEFT";
                else t.swipe_direction = "RIGHT";
            }
            else if (f(t.swipe_finish_y) >= p.i.swipe_finish_y_limit.from && f(t.swipe_finish_x) <= p.i.swipe_finish_x_limit.to) {
                if (t.swipe_finish_y < 0) t.swipe_direction = "UP";
                else t.swipe_direction = "DOWN";
            }
        }
    }
    v = p.i.touches[0];
    p.i.touch_x = v.x;
    p.i.touch_y = v.y;
    p.i.tap = !0;
    p.i.touchdown = !1;
    e.preventDefault();
};

w.ontouchcancel = function(e) {
    for (var i = 0; i < q.length; i++) {
        t = p.i.touches[i];
        t.tap = !1;
        t.touchdown = !1;
    }
    p.i.tap = !1;
    p.i.touchdown = !1;
    e.preventDefault();
};

w.onclick = function() {
    p.i.click = !0;
};

w.onkeydown = function(e) {
    p.i.latest_key_down = e.which || e.keyCode;
};

w.onkeyup = function(e) {
    p.i.latest_key_up = e.which || e.keyCode;
};

w.onwheel = function(e) {
    p.i.wheel_x = e.deltaX;
    p.i.wheel_y = e.deltaY;
    p.i.wheel_z = e.deltaZ;
    
    if (p.i.wheel_x > 0) {
        p.i.wheel_left = !1;
        p.i.wheel_right = !0;
    } else if (p.i.wheel_x < 0) {
        p.i.wheel_left = !0;
        p.i.wheel_right = !1;
    } else if (p.i.wheel_x === 0) p.i.wheel_left = p.i.wheel_right = !1;
    
    if (p.i.wheel_y > 0) {
        p.i.wheel_up = !1;
        p.i.wheel_down = !0;
    } else if (p.i.wheel_y < 0) {
        p.i.wheel_up = !0;
        p.i.wheel_down = !1;
    } else if (p.i.wheel_y === 0) p.i.wheel_up = p.i.wheel_down = !1;
};

p.i.mousedown = function(b) {
    return p.i.latest_mouse_button_down === b;
};

p.i.mouseup = function(b) {
    return p.i.latest_mouse_button_up === b;
};

p.i.swipe = function(d, f) {
    q = p.i.touches[f || 0];
    if (q) return (q.swipe_direction === d);
    else return (p.i.swipe_direction === d);
};

p.i.keydown = function(k) {
    return p.i.latest_key_down === k;
};

p.i.keyup = function(k) {
    return p.i.latest_key_up === k;
};

p.i.cursorState = function(s, i) {
    b.style.height = "100%";
    v = d.getElementsByTagName("html")[0];
    v.style.height = "100%";
    p.canvases[i || 0].style.cursor = s;
    if (p.g.backend === "WebGLRenderingContext" && p.g.ctx2d_enabled) p.g.ctx2d.canvas.style.cursor = s;
    b.style.height = "auto";
    v.style.height = "auto";
};

p.i.hideCursor = function(i) {
    p.i.cursorState("none", i);
};

p.i.showCursor = function(i) {
    p.i.cursorState("auto", i);
};

p.i.setCursor = function(css, i) {
    p.i.cursorState(css, i);
};

p.i.lockPointer = function() {
    f = p.g.ca.requestPointerLock;
    if (f) f();
    if (d.pointerLockElement === p.g.ca && f) f();
};

p.i.unlockPointer = function() {
    d.exitPointerLock();
};

if (p.b.s.GAMEPAD()) {
    if (!w.Windows) n.getGamepads = n.getGamepads || n.webkitGetGamepads || n.webkitGamepads;
    p.i.gamepad_move_horizontal_direction = "";
    p.i.gamepad_move_vertical_direction = "";
    p.i.gamepad_camera_horizontal_direction = "";
    p.i.gamepad_camera_vertical_direction = "";
    p.i.GAMEPAD_ANALOG_UP = -p.i.gamepad_threshold;
    p.i.GAMEPAD_ANALOG_DOWN = p.i.gamepad_threshold;
    p.i.GAMEPAD_ANALOG_LEFT = p.i.gamepad_threshold;
    p.i.GAMEPAD_ANALOG_RIGHT = -p.i.gamepad_threshold;
    p.i.GAMEPAD_MOVE_ANALOG = 1;
    p.i.GAMEPAD_CAMERA_ANALOG = 2;

    if (w.Windows) {
        p.i.GAMEPAD_ANALOG_UP = -p.i.GAMEPAD_ANALOG_UP;
        p.i.GAMEPAD_ANALOG_DOWN = -p.i.GAMEPAD_ANALOG_DOWN;
    }

    p.i.gamepadConnected = function(i) {
        if (w.Windows) return (w.Windows.Gaming.Input.Gamepad.gamepads[i]);
        else return (n.getGamepads()[i]);
    };

    p.i.gamepadID = function(i) {
        if (!w.Windows) {
            v = n.getGamepads()[i];
            if (v) return v.id;
        } else return "XInput STANDARD GAMEPAD";
    };

    p.i.gamepadButtonPressed = function(i, b) {
        if (!w.Windows) {
            v = n.getGamepads()[i];
            if (v) return v.buttons[b].pressed;
        } else {
            if (w.Windows.Gaming.Input.Gamepad.gamepads[i]) {
                v = w.Windows.Gaming.Input.Gamepad.gamepads[i].getCurrentReading();
                if (b === "leftTrigger" || b === "rightTrigger") return (v[b] >= p.i.gamepad_threshold);
                else return ((v.buttons & w.Windows.Gaming.Input.GamepadButtons[b]) != 0);
            }
        }
    };

    p.i.gamepadButtonTouched = function(i, b) {
        if (!w.Windows) {
            v = n.getGamepads()[i];
            if (v) return v.buttons[b].touched;
        } else {
            if (w.Windows.Gaming.Input.Gamepad.gamepads[i]) {
                v = w.Windows.Gaming.Input.Gamepad.gamepads[i].getCurrentReading();
                if (b === "leftTrigger" || b === "rightTrigger") return (v[b] >= p.i.gamepad_threshold);
                else return ((v.buttons & w.Windows.Gaming.Input.GamepadButtons[b]) != 0);
            }
        }
    };

    p.i.gamepadMovement = function(i, a, d) {
        if (!w.Windows) {
            v = n.getGamepads()[i];
            if (v) {
                if (a === p.i.GAMEPAD_MOVE_ANALOG) {
                    if (v.axes[3] <= d) p.i.gamepad_move_vertical_direction = "UP";
                    if (v.axes[3] >= d) p.i.gamepad_move_vertical_direction = "DOWN";
                    if (v.axes[2] <= d) p.i.gamepad_move_horizontal_direction = "LEFT";
                    if (v.axes[2] >= d) p.i.gamepad_move_horizontal_direction = "RIGHT";
                }
                
                if (a === p.i.GAMEPAD_CAMERA_ANALOG) {
                    if (v.axes[1] <= d) p.i.gamepad_camera_vertical_direction = "UP";
                    if (v.axes[1] >= d) p.i.gamepad_camera_vertical_direction = "DOWN";
                    if (v.axes[0] <= d) p.i.gamepad_camera_horizontal_direction = "LEFT";
                    if (v.axes[0] >= d) p.i.gamepad_camera_horizontal_direction = "RIGHT";
                }
            }
        } else {
            v = w.Windows.Gaming.Input.Gamepad.gamepads[i];
            if (v) {
                f = w.Windows.Gaming.Input.Gamepad.gamepads[i].getCurrentReading();
                if (a === p.i.GAMEPAD_MOVE_ANALOG) {
                    if (f.rightThumbstickY >= d) p.i.gamepad_move_vertical_direction = "UP";
                    if (f.rightThumbstickY <= d) p.i.gamepad_move_vertical_direction = "DOWN";
                    if (f.rightThumbstickX <= d) p.i.gamepad_move_horizontal_direction = "LEFT";
                    if (f.rightThumbstickX >= d) p.i.gamepad_move_horizontal_direction = "RIGHT";
                }
                
                if (a === p.i.GAMEPAD_CAMERA_ANALOG) {
                    if (f.leftThumbstickY >= d) p.i.gamepad_camera_vertical_direction = "UP";
                    if (f.leftThumbstickY <= d) p.i.gamepad_camera_vertical_direction = "DOWN";
                    if (f.leftThumbstickX <= d) p.i.gamepad_camera_horizontal_direction = "LEFT";
                    if (f.leftThumbstickX >= d) p.i.gamepad_camera_horizontal_direction = "RIGHT";
                }
            }
        }
    };

    p.i.gamepadAnalogMoved = function(i, a, d) {
        if (d === "UP") q = p.i.GAMEPAD_ANALOG_UP;
        if (d === "DOWN") q = p.i.GAMEPAD_ANALOG_DOWN;
        if (d === "LEFT") q = p.i.GAMEPAD_ANALOG_LEFT;
        if (d === "RIGHT") q = p.i.GAMEPAD_ANALOG_RIGHT;
        p.i.gamepadMovement(i, a, q);
        if (a === p.i.GAMEPAD_MOVE_ANALOG) return (p.i.gamepad_move_horizontal_direction === d || p.i.gamepad_move_vertical_direction === d);
        if (a === p.i.GAMEPAD_CAMERA_ANALOG) return (p.i.gamepad_camera_horizontal_direction === d || p.i.gamepad_camera_vertical_direction === d);
    };
} else {
    p.i.gamepadConnected = function(i) { return !1; };
    p.i.gamepadID = function(i) { return u; };
    p.i.gamepadButtonPressed = function(i, b) { return !1; };
    p.i.gamepadButtonTouched = function(i, b) { return !1; };
    p.i.gamepadMovement = function(i, a, d) { return u; };
    p.i.gamepadAnalogMoved = function(i, a, d) { return !1; };
}

p.i.preventLoop = function () {
    p.i.latest_key_down = -1;
    p.i.latest_key_up = -1;
    p.i.latest_mouse_button_down = -1;
    p.i.latest_mouse_button_up = -1;
    p.i.click = !1;
    p.i.tap = !1;
    p.i.touchdown = !1;
    p.i.wheel_up = !1;
    p.i.wheel_down = !1;
    p.i.wheel_left = !1;
    p.i.wheel_right = !1;
    p.i.swipe_direction = "";
    if (p.i.touches.length > 0) {
        for (var i = 0; i < p.i.touches.length; i++) p.i.touches[i].swipe_direction = "";
    }
    p.i.gamepad_move_horizontal_direction = "";
    p.i.gamepad_move_vertical_direction = "";
    p.i.gamepad_camera_horizontal_direction = "";
    p.i.gamepad_camera_vertical_direction = "";
};

p.i.key = {
    A: 65, B: 66, C: 67, D: 68, E: 69, F: 70, G: 71, H: 72, I: 73, J: 74, K: 75, L: 76, M: 77,
    N: 78, O: 79, P: 80, Q: 81, R: 82, S: 83, T: 84, U: 85, V: 86, W: 87, X: 88, Y: 89, Z: 90,
    ZERO: 48, ONE: 49, TWO: 50, THREE: 51, FOUR: 52, FIVE: 53, SIX: 54, SEVEN: 55, EIGHT: 56,
    NINE: 57, UP: 38, DOWN: 40, LEFT: 37, RIGHT: 39, SPACE: 32, TAB: 9, SHIFT: 16, CONTROL: 17,
    ALT: 18, BACKSPACE: 8, ENTER: 13, NUMLOCK: 144, OS: 91, UNIDENTIFIED: 0, HOME: 36, PGUP: 33,
    PGDN: 34, CLEAR: 12, DELETE: 46, ESCAPE: 27, INSERT: 45
};

p.i.tvkey = {
    UP: 38, DOWN: 40, LEFT: 37, RIGHT: 39, CHANNEL_UP: 516, CHANNEL_DOWN: 517, CONTEXT: 623, RED: 588,
    GREEN: 589, YELLOW: 590, BLUE: 591, ENTER: 13, INFO: 615, ASPECT: 642, LASTVIEW: 651, ZERO: 48, ONE: 49, TWO: 50,
    THREE: 51, FOUR: 52, FIVE: 53, SIX: 54, SEVEN: 55, EIGHT: 56, NINE: 57, BACK: 8, RETURN: 8, PLAY: 250, PAUSE: 19,
    STOP: 178, RECORD: 603, FORWARD: 228, REWIND: 227, FAST_FORWARD: 176, REPLAY: 177
};

p.i.button = {
    LEFT_MOUSE_BUTTON: 0, RIGHT_MOUSE_BUTTON: 2, MIDDLE_MOUSE_BUTTON: 1, A: 0, B: 1, XBOX_X: 2,
    Y: 3, LB: 4, RB: 5, LT: 6, RT: 7, SELECT: 8, BACK: 8, VIEW: 8, START: 9, MENU: 9, LEFT_ANALOG_STICK: 10, RIGHT_ANALOG_STICK: 11,
    UP: 12, DOWN: 13, LEFT: 14, RIGHT: 15, PLAYSTATION_X: 0, O: 1, SQUARE: 2, TRIANGLE: 3, L1: 4, R1: 5, L2: 6, R2: 7
};

if (w.Windows) {
    p.i.button.A = "a";
    p.i.button.B = "b";
    p.i.button.XBOX_X = "x";
    p.i.button.Y = "y";
    p.i.button.UP = "dpadUp";
    p.i.button.DOWN = "dpadDown";
    p.i.button.LEFT = "dpadLeft";
    p.i.button.RIGHT = "dpadRight";
    p.i.button.LB = "leftShoulder";
    p.i.button.RB = "rightShoulder";
    p.i.button.LT = "leftTrigger";
    p.i.button.RT = "rightTrigger";
    p.i.button.BACK = p.i.button.SELECT = p.i.button.VIEW = "view";
    p.i.button.START = p.i.button.MENU = "menu";
    p.i.button.LEFT_ANALOG_STICK = "leftThumbstick";
    p.i.button.RIGHT_ANALOG_STICK = "rightThumbstick";
}
